Add happens query for tasks

Kongolan opened the pull request:
### What it does?
It's a new query string, which searchs for all the date types: _starting, scheduled, due_ and outputs it. 

### Why is it needed?
Currently you can't query just by date, due, starting and scheduled are all seperated, so you need multiple tasks queries. I think it will be helpful to just see what is coming tomorrow from all sorts of date types set to that date.

### Example Usage
```tasks
not done
happens on tomorrow
```
### Features
It support the "after" and "before" query as well and it will respect start dates first, in case of multiple date types added to the task.
```tasks
not done
happens before today
```

If you don't want a certain date time shown, you can use the default filter:
`no scheduled date`

```tasks
not done
no scheduled date
happens after today
happens before in 1 week
```

schemar commented:
Thank you @Kongolan! Your rationale makes absolute sense. Thank you for the PR and the good description :heart: 

Just one thought regarding the implementation:
Right now, the implementation makes the (reasonable) assumption, that I only want to see the task on the earliest occurrence of `start`, `scheduled`, or `due`.

What do you think about the following alternative:
`happens` is true whenever one of the dates matches? In other words, even if the task has a start date in the past, if it has a matching scheduled date it shows again for that match.

Let me know if my description is unclear :see_no_evil:

Kongolan commented:
@schemar I added your suggestion in my latest commit. I also rebased to the newest updates, to make merging easier.

### How it will behave now
`happens` should now be indifferent of which date is matching. Any match will return true.

### But there is a small problem as well 
If we use `happens after today` and another query with `happens on today` and we also have `start` set on `today` and another date like `due` set on `tomorrow` then it will show in both lists, because it's starting today, but it's also upcoming soon.

### Hint
Can you review if the indention is proper for you, since my formatter seems to handle things differently than yours?

schemar commented:
> @schemar I added your suggestion in my latest commit. I also rebased to the newest updates, to make merging easier.
> ### How it will behave now
> 
> `happens` should now be indifferent of which date is matching. Any match will return true.
> ### But there is a small problem as well
> 
> If we use `happens after today` and another query with `happens on today` and we also have `start` set on `today` and another date like `due` set on `tomorrow` then it will show in both lists, because it's starting today, but it's also upcoming soon.
> ### Hint
> 
> Can you review if the indention is proper for you, since my formatter seems to handle things differently than yours? Also I'm not sure, why i couldn't give `date` a `: Moment` type in the arrow function, it just didn't work. So maybe add a type safety there.

Thank you so much @Kongolan! :green_heart:  I think it's mergeable as soon as I find the time to merge it proper :+1: 

Regarding your "small problem": I think that's the logical behavior and in my opinion makes sense. It does "happen" in both cases.

schemar reviewed (approved):
Lovely, thank you so much!

claremacrae commented:
This looks really useful! There's some quite serious logic here, it would be really good to see some test cases to exercise the logic - to guard against future accidental breakage.

claremacrae commented:
> This looks really useful! There's some quite serious logic here, it would be really good to see some test cases to exercise the logic - to guard against future accidental breakage.

I'd be happy to pair with anyone willing to work on this, if that would help...

schemar commented:
> This looks really useful! There's some quite serious logic here, it would be really good to see some test cases to exercise the logic - to guard against future accidental breakage.

I totally agree. I tried adding tests, but gave up as it wasn't worth the time. It's not that easy, unfortunately. You need to tell Jest to run the test as a "browser" in order to have `window` available. That's easy, you can copy/paste how to do that from the `Sort` test, for example:

```ts
/**
 * @jest-environment jsdom
 */
import moment from 'moment';

window.moment = moment;
```

However, when running the tests it returned the error that `chrono` was `undefined` at [line 525](https://github.com/schemar/obsidian-tasks/blob/859adc6eae8b66b90ef4a36d52cca3f44c4b425c/src/Query.ts#L525), even though I `import chrono from 'chrono-node'` at the to of the `Query` file. I wasn't able to fix that within a reasonable amount of time. Maybe someone else knows what I did wrong?

claremacrae commented:
OK brilliant - I now have a good example for my Community Talk, to show how to make this testable and add tests! :-)

claremacrae commented:
> OK brilliant - I now have a good example for my Community Talk, to show how to make this testable and add tests! :-)

I hope!

claremacrae commented:
> You need to tell Jest to run the test as a "browser" in order to have `window` available. That's easy, you can copy/paste how to do that from the `Sort` test, for example:
> 
> ```ts
> /**
>  * @jest-environment jsdom
>  */
> import moment from 'moment';
> 
> window.moment = moment;
> ```

That was a really useful tip. Thank you.

> However, when running the tests it returned the error that `chrono` was `undefined` at [line 525](https://github.com/schemar/obsidian-tasks/blob/859adc6eae8b66b90ef4a36d52cca3f44c4b425c/src/Query.ts#L525), even though I `import chrono from 'chrono-node'` at the to of the `Query` file. I wasn't able to fix that within a reasonable amount of time. Maybe someone else knows what I did wrong?

I was able to reproduce that exact same behaviour, and after a lot of googling, eventually found that the [chrono-node docs](https://www.npmjs.com/package/chrono-node) suggest this style of import:

```ts
import * as chrono from 'chrono-node';
```

When I add that moment code from above, and switch to that style of import in `Query.test.ts` the following code executes fine:

```ts
        it('can create a happens filter', () => {
            // Arrange
            const input2 = 'happens before 2022-02-03';
            const query = new Query({ source: input2 });

            // Assert
            // TODO
            expect(query.error).toBeUndefined();
        });
```

claremacrae commented:
This was included in the 1.5.0 release.
